src/charm.py

```diff
@@ -723,6 +775,42 @@ def _on_run_migration_action(self, event: ActionEvent) -> None:
             return
         event.log("Successfully migrated the database.")
 
+    def _on_loki_consumer_endpoint_joined(self, event: LogProxyEndpointJoined) -> None:
+        logger.info("Loki Push API endpoint joined")
+
+    def _on_loki_consumer_endpoint_departed(self, event: LogProxyEndpointDeparted) -> None:
+        logger.info("Loki Push API endpoint departed")
+
+    def _promtail_error(self, event: PromtailDigestError):
+        logger.error(event.message)
+        self.unit.status = BlockedStatus(event.message)
+
+    def _on_update_status_handle_logging(self, event: HookEvent) -> None:
+        if not self._container.can_connect():
+            logger.info(
+                "Cannot connect to Kratos container. Log file check delayed to next status update."
+            )
+            return
+
+        if not self.unit.status == ActiveStatus():
+            logger.info(
+                "Kratos unit currently inactive. Log file check delayed to next status update."
+            )
+            return
+
+        log_file_info_list = self._container.list_files(path=self._log_path)
+
+        if not log_file_info_list:
+            logger.info(
+                "Log file has not been created. Log file check delayed to next status update."
+            )
+            return
+
+        if int(log_file_info_list[0].size) > self.config["log_max_size"] * 1000000:
+            logger.info("Log file greater than allowed size. Redeploying container.")
```

Comment:
**Question:** What do you mean by "redeploying container"?

Perhaps [head-tail](https://unix.stackexchange.com/questions/367055/how-do-i-trim-bytes-from-the-beginning-and-end-of-a-file) or [RotatingFileHandler](https://stackoverflow.com/questions/9106795/python-logging-and-rotating-files) could be useful?

Reply:
I realize since that this function only restarts the service, not the container. I'm not sure how I can use either solution without introducing some problems with filehandlers or a race condition for the log file, therefore for now I'll remove the log rotation functionality all together.

Reply:
Haven't checked, but how about
1. stop service
2. trim log
3. start service

BTW what does kratos do with logs? Just let them grow?

Reply:
Kratos just writes to stdout, I suppose in that case log rotation is managed by kubelet. We discussed with @bencekov a similar approach to the one you described (not trimming the logs, but rotating the log file). 

Tbh I am not familiar with promtail and I don't know how it would behave if we trimmed logs from the file, but it sounds error prone. Do you think this is the best way forward or are there any other projects that have implemented this approach?

Reply:
Ah right.
I think promtail is able to pick up a head-trim, but @rbarry82 can probably advise more accurately.
Grafana labs [recommend](https://grafana.com/docs/loki/latest/clients/promtail/logrotation/#rename-and-create) using "rename and create".